### PR TITLE
ginkgo: add v1.8.0

### PR DESCRIPTION
**Pull request by MarcelKoch**
This is a continuation of https://github.com/spack/spack/pull/44923. I take over for @tcojean. This should contain a fix for the CI failures in the old PR.

@tcojean, @hartwiganzt one of you might want to close the old PR.

### CODE REVIEW

File: var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch
Comment (wdconinc) [inline var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch:1]: Could we pull this in as a url when https://github.com/ginkgo-project/ginkgo/pull/1668 is merged? Keeps us from having to store local copies of patches.

Comment (MarcelKoch) [reply var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch:1]: Sure, since the CI passed this time, I would consider the PR to be complete.

**Review by tcojean — APPROVED** [review]
LGTM, thanks Marcel!

**Review by balay — APPROVED** [review]

File: var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch
Comment (balay) [reply var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch:1]: Just to add to this - perhaps its good if the ginkgo PR is merged first - and then the merge commit is listed here [instead of the diff from the PR].

But I guess thats not a blocker [as I see similar PR diffs listed in current spack sources]

Comment (wdconinc) [reply var/spack/repos/builtin/packages/ginkgo/find_rocthrust.patch:1]: A merge PR is fine, as long as it is stable.

The reason for waiting here until it is merged upstream is to avoid that someone upstream goes ahead and rebases the PR there right before merging it. That would make the hash here invalid.

File: var/spack/repos/builtin/packages/ginkgo/package.py [hunk 1]
```diff
@@ -18,14 +18,15 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
 
     test_requires_compiler = True
 
-    maintainers("tcojean", "hartwiganzt")
+    maintainers("MarcelKoch", "hartwiganzt")
 
     tags = ["e4s"]
 
-    license("MIT")
+    license("BSD-3-Clause")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/ginkgo/package.py:25]: Is the MIT license still valid for older versions?  If so, you can add the `when` argument to indicate when each license applies.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Confirmed the commit, rest LGTM except have question about the license change.

**Review by eugeneswalker — APPROVED** [review]

File: var/spack/repos/builtin/packages/ginkgo/package.py (hunk 1 above)
Comment (MarcelKoch) [reply var/spack/repos/builtin/packages/ginkgo/package.py:25]: The first released license: https://github.com/ginkgo-project/ginkgo/blob/v1.0.0/LICENSE is already BSD-3-clause, so all Ginkgo versions are under this license.

**Review by tldahlgren — APPROVED** [review]

File: var/spack/repos/builtin/packages/ginkgo/package.py [hunk 2]
```diff
@@ -57,10 +58,11 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("cmake@3.13:", type="build", when="@1.4.0:1.6.0")
     depends_on("cmake@3.16:", type="build", when="@1.7.0:")
     depends_on("cmake@3.18:", type="build", when="+cuda@1.7.0:")
+    depends_on("cmake@3.21:", type="build", when="+rocm@1.8.0:")
     depends_on("cuda@9:", when="+cuda @:1.4.0")
     depends_on("cuda@9.2:", when="+cuda @1.5.0:")
     depends_on("cuda@10.1:", when="+cuda @1.7.0:")
-    depends_on("mpi", when="+mpi")
+    depends_on("mpi@3.1:", when="+mpi")
```
Comment (balay) [inline var/spack/repos/builtin/packages/ginkgo/package.py:65]: I'm seeing a strange issue with this change. An old version of openmpi is getting picked up. I'm not sure I completely understand this.

```console
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ ./bin/spack spec ginkgo ^openmpi |grep openmpi@
 -       ^openmpi@1.7.4%gcc@11.4.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~internal-libevent~internal-pmix~java~lustre~memchecker~openshmem~orterunprefix~romio+rsh~singularity~sqlite3~static~thread_multiple+vt+wrapper-rpath build_system=autotools fabrics=none romio-filesystem=none schedulers=none arch=linux-ubuntu22.04-zen4
```
However - if I revert this change - the correct/latest version gets used
```console
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ git diff
diff --git a/var/spack/repos/builtin/packages/ginkgo/package.py b/var/spack/repos/builtin/packages/ginkgo/package.py
index b4638d9666..0c1f6a5089 100644
--- a/var/spack/repos/builtin/packages/ginkgo/package.py
+++ b/var/spack/repos/builtin/packages/ginkgo/package.py
@@ -62,7 +62,7 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("cuda@9:", when="+cuda @:1.4.0")
     depends_on("cuda@9.2:", when="+cuda @1.5.0:")
     depends_on("cuda@10.1:", when="+cuda @1.7.0:")
-    depends_on("mpi@3.1:", when="+mpi")
+    depends_on("mpi", when="+mpi")
 
     depends_on("rocthrust", when="+rocm")
     depends_on("hipsparse", when="+rocm")
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ ./bin/spack spec ginkgo ^openmpi |grep openmpi@
 -       ^openmpi@5.0.5%gcc@11.4.0~atomics~cuda~gpfs~internal-hwloc~internal-libevent~internal-pmix~java~lustre~memchecker~openshmem~romio+rsh~static+vt+wrapper-rpath build_system=autotools fabrics=none romio-filesystem=none schedulers=none arch=linux-ubuntu22.04-zen4
```

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/ginkgo/package.py:65]: @alalazo Do you know why this might be happening?

Comment (balay) [reply var/spack/repos/builtin/packages/ginkgo/package.py:65]: Don't know the exact mpi version to openmpi version map,  but changing the format to the one used in mpich gets this working.
```console
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ git diff
diff --git a/var/spack/repos/builtin/packages/openmpi/package.py b/var/spack/repos/builtin/packages/openmpi/package.py
index 22bc36c360..8560519c93 100644
--- a/var/spack/repos/builtin/packages/openmpi/package.py
+++ b/var/spack/repos/builtin/packages/openmpi/package.py
@@ -570,10 +570,10 @@ class Openmpi(AutotoolsPackage, CudaPackage):
     variant("internal-libevent", default=False, description="Use internal libevent")
     variant("openshmem", default=False, description="Enable building OpenSHMEM")
 
-    provides("mpi")
-    provides("mpi@:2.2", when="@1.6.5")
-    provides("mpi@:3.0", when="@1.7.5:")
-    provides("mpi@:3.1", when="@2.0.0:")
+    provides("mpi@:4.0")
+    provides("mpi@:3.1", when="@:4.1")
+    provides("mpi@:3.0", when="@:1.10")
+    provides("mpi@:2.2", when="@:1.7")
 
     if sys.platform != "darwin":
         depends_on("numactl")
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ ./bin/spack spec ginkgo ^openmpi |grep openmpi@
 -       ^openmpi@5.0.5%gcc@11.4.0~atomics~cuda~gpfs~internal-hwloc~internal-libevent~internal-pmix~java~lustre~memchecker~openshmem~romio+rsh~static+vt+wrapper-rpath build_system=autotools fabrics=none romio-filesystem=none schedulers=none arch=linux-ubuntu22.04-x86_64
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack$ 
```

Comment (balay) [reply var/spack/repos/builtin/packages/ginkgo/package.py:65]: Updated fix at https://github.com/spack/spack/pull/46102

Comment (balay) [reply var/spack/repos/builtin/packages/ginkgo/package.py:65]: Since #46102 is having issues - hoping  we can use https://github.com/spack/spack/pull/46292 until these issues are resolved.
